pgjdbc/src/main/java/org/postgresql/jdbc/PgResultSet.java

```diff
@@ -2232,12 +2250,28 @@ public boolean next() throws SQLException {
       }
 
       // Execute the fetch and update this resultset.
-      connection.getQueryExecutor()
-          .fetch(cursor, new CursorResultHandler(), fetchRows, adaptiveFetch);
+      if (refCursorName == null || refCursorName.isEmpty()) {
+        castNonNull(resultCursor, "resultCursor");
+        connection.getQueryExecutor()
+            .fetch(resultCursor, new CursorResultHandler(), fetchRows, adaptiveFetch);
+        // .fetch(...) could update this.cursor, and cursor==null means
+        // there are no more rows to fetch
 
-      // .fetch(...) could update this.cursor, and cursor==null means
-      // there are no more rows to fetch
-      closeRefCursor();
+      } else {
+        StringBuilder sb = new StringBuilder("FETCH FORWARD ");
+        sb.append(fetchRows);
+        sb.append(" IN ");
+        castNonNull(refCursorName, "refCursorName");
+        Utils.escapeIdentifier(sb, refCursorName);
+        final Query cursorForward = connection.getQueryExecutor().createSimpleQuery(sb.toString());
+        // pass maxRows=0 since fetchSize has already been corrected to account for maxRows
+        connection.getQueryExecutor().execute(cursorForward, null, new CursorResultHandler(), 0, fetchSize, QueryExecutor.QUERY_ONESHOT | QueryExecutor.QUERY_SUPPRESS_BEGIN
+            | QueryExecutor.QUERY_EXECUTE_AS_SIMPLE );
```

Comment:
I wonder if we can reuse the existing "extended protocol" fetch to fetch the data rather than build "fetch forward" commands dynamically